exchanges/stream/websocket.go

```diff
@@ -434,6 +434,13 @@ func (w *Websocket) connect() error {
 			break
 		}
 
+		if len(subs) != 0 {
+			if err = w.AddSuccessfulSubscriptions(conn, subs...); err != nil {
+				multiConnectFatalError = fmt.Errorf("%v Error adding successful subscriptions %w", w.exchangeName, err)
+				break
+			}
+		}
+
```

Comment:
I don't think we can do this all-or-nothing approach.
I'm not sure we should bail on connect in *either* situation (multi or not), because failing to subscribe does not mean failing to connect. We should probably log the error and continue on.

But in this instance, if we managed to subscribe to 4 subs out of 5, we don't want to have not recorded that we subscribed to them.

As Subscriber is right now, I don't think we can delay Adding until here. It has to be in the manageSubs with more granular visibility.

This feels like it's raised a new issue for me of "connect shouldn't error if some subs fail", but that's separate.

Reply:
Yeah I differ in perspective on that, I regard a failure on 1 is a failure on the entire set. Especially when trading I want to ensure when I connect its subscribed to all the required subscriptions I want. 

If there is a failure on subscription its either a malformed outbound payload or the response from the connection is delayed in both cases the connection is not ideal for trading ops. A retry under a for loop is easier and less complicated after handling error from `connect` if it is a subscription delay error.

Reply:
Feel free to return error from connect.
I don't agree with it, but sure.

_Aside: The best solution to my problem with connect returning error is to not have connect do other things, and have `Run` call connect, then auth, then subscribe. Clearer responsibilities for methods.
I really can't get onboard with calling a method called connect, it connects, then does something else and returns an error about that other thing._

But for today, my problem is: Subs succeeded, but you haven't added any of them as successful.

If you want to return error from connect on sub failure, please ensure the successful ones are added first, then return error.

Reply:
> Aside: The best solution to my problem with connect returning error is to not have connect do other things, and have Run call connect, then auth, then subscribe. Clearer responsibilities for methods.
I really can't get onboard with calling a method called connect, it connects, then does something else and returns an error about that other thing.

Well no the current design differs slightly from  `connect -> auth -> subscribe` so we retrieve subscriptions -> only if subscriptions are available attempt connection/connect -> authenticate connection if available -> then subscribe (auth/unauth subs). We don't need to spawn connection/connect when there are no subscriptions, there is a caveat here for when this is allowed for a dedicated outbound connection for requests. This allows us to scale in the event of connection only allowing N subs per connection in future as well (OKX). Connect() means nothing without subscriptions. I guess we could rename to something different? 

> But for today, my problem is: Subs succeeded, but you haven't added any of them as successful.
If you want to return error from connect on sub failure, please ensure the successful ones are added first, then return error.

I don't know what you mean by this sorry? In regards to the current state of the PR I have added in `AddSuccessfulSubscriptions` but we are not going to do this any way? If this is directly related to `I will investigate is check the connection subscription status after subscriptions to see if they have been added or removed. So that at least it complains that it needs to be added in the exchange packages themselves.` If there are issues yeah I will address that for this. 

I just want to steer this into a less complicated scenario for exchange implementations, pulling everything apart and toiling on things like this, seems like a good to have instead of a need to have. Everything you have said on your concerns has been on point. I am just trying to give a different perspective.

Reply:
> Connect() means nothing without subscriptions.

We have non-subscription ws methods sometimes, right?
Or do you mean purely in the context of GateIO?

Reply:
> I don't know what you mean by this sorry?

If this PR merges as it currently is, it means that subs which succeeded are not added as successful, if any failed.

Previously they were added as they succeeded by the exchange, one by one.
Now, if one of them fails, then L434 `break` means none of them is added, and that's wrong.

I don't know how you can do this outside of the exchange implementation, because it's so varied when and how we decide it's successful.